### PR TITLE
Consul acl switch

### PR DESCRIPTION
**Pull request by zhenik**

### CODE REVIEW

**Comment by zhenik** [comment]
I added a diagram in PNG format, I suppose that it should be a mermaid or some code format. WDYT @fredrikhgrelland ?
life-cycle

File: ansible/bootstrap.yml [hunk 1]
```diff
@@ -11,10 +11,49 @@
       include_tasks: "{{ item }}"
       with_fileglob: "/vagrant/vagrant/conf/ansible/playbooks/prestart/*"
 
+    - name: Verify default variables
+      include_tasks: default_vars.yml
+
+    - name: Verify Consul ACL default policy
+      template: src=/etc/ansible/templates/consul-acl.hcl.j2 dest=/etc/consul.d/config-merge1.hcl
```
Comment (fredrikhgrelland) [inline ansible/bootstrap.yml:18]: Why not 
```suggestion
      template: src=/etc/ansible/templates/consul-acl.hcl.j2 dest=/etc/consul.d/config-acl.hcl
```

File: ansible/bootstrap.yml [hunk 2]
```diff
@@ -11,10 +11,49 @@
...
+      retries: 5
+      delay: 1
+      when: consul_agent_acl_default_policy == "deny"
+
+    - name: Verify Consul ACL deny policy - access granted with correct token
+      uri:
+        url: http://127.0.0.1:8500/v1/acl/tokens
+        status_code: 200
+        headers:
+          X-Consul-Token: master
```
Comment (fredrikhgrelland) [inline ansible/bootstrap.yml:40]: Make this to a variable

File: template/README.md [hunk 3]
```diff
@@ -21,6 +21,35 @@ You may change the hashistack configuration or add aditional pre and post steps
...
+Current supported variations:
+```text
+* ACL default policy for consul (deny or allow)
+```
+To choose other config variation you need to override ansible variables. There are two options how to override ansible variables:
+* First option via `ANSIBLE_AGRS`
+```bash
+ANSIBLE_ARGS='--extra-vars "consul_agent_acl_default_policy=deny"' vagrant up --provision
+```
+* Second option via prestart, adding fact in `vagrant/conf/ansible/playbooks/prestart/0-example.yml`
```
Comment (fredrikhgrelland) [inline template/README.md:35]:
```suggestion
* Second option via prestart, adding fact in `vagrant/conf/ansible/playbooks/prestart/0-acl.yml`
```

Comment (fredrikhgrelland) [reply template/README.md:35]: What do you think?

File: template/README.md [hunk 4]
```diff
@@ -21,6 +21,35 @@ You may change the hashistack configuration or add aditional pre and post steps
...
+
 ### Pre- and post-startup ansible playbooks
 This vagrant box will execute ansible playbooks put in two special directories [vagrant/conf/ansible/playbooks/prestart](vagrant/conf/ansible/playbooks/prestart) and [vagrant/conf/ansible/playbooks/poststart](vagrant/conf/ansible/playbooks/poststart). These playbooks will be executed before and after the box's bundled startup sequence, respectively. This gives the flexibility to configure all aspects of the hashistack as well as run tasks needed for tests or demo purposes as part of `vagrant up` Note; The playbooks are included into the main run, so the syntax in the [example](vagrant/conf/ansible/playbooks/prestart/0-example.yml) must be followed..  
 They will be run in lexical order, and prefixing with numbers is a good way to get the order you want.
+
+### Vagrant box life-cycle
+![img](../docs/life-cycle.png)
+1. `install.yml` - installing required software
+2. `default_vars.yml` - setup default ansible variables
+3. `prestart/0-example.yml` - prestart scripts, running before hashistack software will start
```
Comment (fredrikhgrelland) [inline template/README.md:52]:
```suggestion
3. `prestart/*.yml` - prestart scripts, running before hashistack software will start
```

File: template/README.md [hunk 5]
```diff
@@ -21,6 +21,35 @@ You may change the hashistack configuration or add aditional pre and post steps
...
+4. `bootstrap.yml` - verify ansible variables and software configuration, run hashistack software & verify that it started correctly
+5. `poststart/0-example.yml` - poststart scripts, running after hasistack software runs and ready
```
Comment (fredrikhgrelland) [inline template/README.md:54]:
```suggestion
5. `poststart/*.yml` - poststart scripts, running after hasistack software runs and ready
```

File: ansible/bootstrap.yml (hunk 2 above)
Comment (zhenik) [reply ansible/bootstrap.yml:40]: Do you mean token (value of `X-Consul-Token` header)?

File: template/README.md (hunk 3 above)
Comment (zhenik) [reply template/README.md:35]: I think we will have several extra variables that will be NOT necessary associated with ACL functionality. So, it is the user's choice of how to name prestart `yml` script file. Considering this concrete example, I would suggest having one prestart script file `0-example.yml` which will contain an example of how to override extra vars (facts)

File: ansible/bootstrap.yml (hunk 2 above)
Comment (fredrikhgrelland) [reply ansible/bootstrap.yml:40]: Well-- "master" should be replaced with a variable in the playbook with the value of "master"

File: template/README.md (hunk 3 above)
Comment (fredrikhgrelland) [reply template/README.md:35]: I disagree. Yes, we will have many features. 0-example.yml gives no visual indication on what it is. However 0-acl.yml and 0-enterprise.yml etc wil definitly have that impact. The files can be commented by default. All you need to do is uncomment them? or they may reside in a sub-folder ( examples ?) ready to go?

Comment (zhenik) [reply template/README.md:35]: Would you like that I create `vagrant/conf/ansible/playbooks/prestart/0-acl.yml` with commented example? I would also bring the user's attention on the prestart script sequence and maybe name it `1-acl.yml`, WDYT?

File: ansible/bootstrap.yml (hunk 2 above)
Comment (zhenik) [reply ansible/bootstrap.yml:40]: I am thinking about additional functionality, be able to override consul acl master token value, via extra_vars. Do we need it or it does not make sense, due that this vagrant box is not for production usage?

**Comment by fredrikhgrelland** [comment]
> I added a diagram in PNG format, I suppose that it should be a `mermaid diagram` or some other code format diagram. WDYT @fredrikhgrelland ?
> life-cycle

mermaid diagram is cool, but as long as the image is reproducible from the repo it is ok

File: template/README.md (hunk 3 above)
Comment (fredrikhgrelland) [reply template/README.md:35]: Yes.

File: ansible/bootstrap.yml (hunk 2 above)
Comment (fredrikhgrelland) [reply ansible/bootstrap.yml:40]: Not right now. variables make it easy to change and code clean. DRY

Comment (zhenik) [reply ansible/bootstrap.yml]: done

File: template/README.md (hunk 3 above)
Comment (zhenik) [reply template/README.md:35]: done

**Pull request merged**